modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyAllocator.java

```diff
@@ -353,4 +361,63 @@ public ByteBufAllocator getDelegate() {
             return delegate;
         }
     }
+
+    static class TrashingByteBuf extends WrappedByteBuf {
+
+        private volatile boolean trashed = false;
+
+        protected TrashingByteBuf(ByteBuf buf) {
+            super(buf);
+        }
+
+        @Override
+        public boolean release() {
+            if (refCnt() == 1) {
+                trashContent();
+            }
```

Comment:
Hmm I'm kind of suspicious about the thread-safety here. Someone else could in theory call `retain()` after we checked `refCnt()` (and wiped out the content) but before we called `super.release()`, and then the other caller would think they have a valid ref to a buffer whose contents we just trashed.

OTOH I suppose that would be invalid with our expected usage pattern, we should never be trying to acquire a ref to a potentially-dead buffer. Maybe we should `assert trashed == false` in `retain()`?

A comment about this would be nice too.

Reply:
Will add assertion and comment. Overall, I dont agree with direction of using unsafe byte arrays behind reference counters and whole family of problems it brings. We should remove access to byte arrays from the http content.

Reply:
> Maybe we should assert trashed == false in retain()?

It's not valid to retain buffer when count reach 0, netty will throw runtime exception. We should see exception now when it happens.

Reply:
Yeah it's not valid but it's also not obvious from the code that this isn't valid. Suggest a comment like this:

```diff
diff --git a/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyAllocator.java b/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyAllocator.java
index 7fd092c1964..791d55481d5 100644
--- a/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyAllocator.java
+++ b/modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/NettyAllocator.java
@@ -373,6 +373,7 @@ public class NettyAllocator {
         @Override
         public boolean release() {
             if (refCnt() == 1) {
+                // see [NOTE on racy trashContent() calls]
                 trashContent();
             }
             return super.release();
@@ -381,11 +382,20 @@ public class NettyAllocator {
         @Override
         public boolean release(int decrement) {
             if (refCnt() == decrement && refCnt() > 0) {
+                // see [NOTE on racy trashContent() calls]
                 trashContent();
             }
             return super.release(decrement);
         }

+        // [NOTE on racy trashContent() calls]: We trash the buffer content _before_ reducing the ref count to zero, which looks racy
+        // because in principle a concurrent caller could come along and successfully retain() this buffer to keep it alive after it's been
+        // trashed. Such a caller would sometimes get an IllegalReferenceCountException ofc but that's something it could handle - see for
+        // instance org.elasticsearch.transport.netty4.Netty4Utils.ByteBufRefCounted.tryIncRef. Yet in practice this should never happen,
+        // we only ever retain() these buffers while we know them to be alive (i.e. via RefCounted#mustIncRef or its moral equivalents) so
+        // it'd be a bug for a caller to retain() a buffer whose refcount is heading to zero and whose contents we've already decided to
+        // trash.
+
         private void trashContent() {
             if (trashed == false) {
                 trashed = true;
```